allennlp/data/dataset_readers/wikitables/type_declaration.py

```diff
@@ -0,0 +1,447 @@
+
+    Note that ANY_TYPE in NLTK's type system doesn't work like a wild card. Once the type of a variable gets
+    resolved to a specific type, NLTK changes the type of that variable to that specific type. Hence, what
+    NLTK calls "ANY_TYPE", is essentially a "yet-to-be-decided type". This is a problem because we may want the
+    same variable to bind to different types within a logical form, and using ANY_TYPE for this purpose will
+    cause a resolution failure. For example the count function may apply to both rows and cells in the same
+    logical form, and making count of type ``ComplexType(ANY_TYPE, DATE_NUM_TYPE)`` will cause a resolution
+    error. This class lets you define ``ComplexType``s with placeholders that are actually wild cards.
+
+    The subclasses of this abstract class need to do two things:
+        1) Override the property ``_signature`` to define the type signature (this is just the signature's
+        string representation and will not affect type inference or checking). You will see this signature in
+        action sequences.
+        2) Override ``resolve`` to resolve the type appropriately (see the docstring in ``resolve`` for more
+        information).
+    """
+    @property
+    def _signature(self):
+        raise NotImplementedError
+
+    @overrides
+    def resolve(self, other):
+        """
+        This method is central to type inference and checking. When a variable's type is being checked, we
+        compare what we know of its type against what is expected of its type by its context. The expectation
+        is provided as ``other``. We make sure that there are no contradictions between this type and other,
+        and return an updated type which may be more specific than the original type.
+
+        For example, say this type is of the function variable F in F(cell), and we start out with <?, d> (that
+        is, it takes any type and returns d). Now we have already resolved `cell` to be of type `e`. Then
+        ``resolve`` gets called with other = <e, ?>, because we know F is a function that took a constant of
+        type `e`. When we resolve <e, ?> against <?, d>, there will not be a contradiction, because any type
+        can be successfully resolved against ?. Finally we return <e, d> as the resolved type.
+
+        As a counter example, if we are trying to resolve <?, d> against <?, e>, the resolution fails, and in
+        that case, this method returns ``None``.
+
+        Note that resolution may be unidirectional because of ?, and so in the subclasses of this type, we
+        explicitly resolve in both directions.
+        """
+        raise NotImplementedError
+
+    @overrides
+    def __eq__(self, other):
+        return self.__class__ == other.__class__
+
+    @overrides
+    def matches(self, other):
+        return self == other or self == ANY_TYPE or other == ANY_TYPE
+
+    @overrides
+    def __str__(self):
+        if self == ANY_TYPE:
```

Comment:
Same as above. This is because `ANY_TYPE.first = ANY_TYPE` and `ANY_TYPE.second = ANY_TYPE` by definition, and `ComplexType` checks the two fields in its overridden `_eq_` method.
The point of this comparison here is to return `?` (`ANY_TYPE._str_()`) if the type remains unresolved.